### ISSUE TITLE
关于DeepFM中的数值特征处理

### ISSUE THREAD

**Issue by timor1988** [EternalStarICe/recommendation-system-model#1]
你好。
代码中FM部分直接去掉了13个数值特征，只使用了26类别特征。然后DNN部分加上了数值特征。
论文提供的模型结构图里面，FM和DNN共享Dense Embeddings。这样FM和DNN的输入相比，少了数值特征。
在下面这篇文章的代码里，其把数值特征也映射到了4维的特征空间。
https://zhuanlan.zhihu.com/p/84526966
请问，在DeepFM里面，对数值特征的处理，有什么思路或者技巧么？

**Comment by EternalStarICe** [comment]
你好！

**Comment by HanpyBin** [comment]
> 你好。
> 代码中FM部分直接去掉了13个数值特征，只使用了26类别特征。然后DNN部分加上了数值特征。
> 论文提供的模型结构图里面，FM和DNN共享Dense Embeddings。这样FM和DNN的输入相比，少了数值特征。
> 在下面这篇文章的代码里，其把数值特征也映射到了4维的特征空间。
> https://zhuanlan.zhihu.com/p/84526966
> 请问，在DeepFM里面，对数值特征的处理，有什么思路或者技巧么？



> 你好。
> 代码中FM部分直接去掉了13个数值特征，只使用了26类别特征。然后DNN部分加上了数值特征。
> 论文提供的模型结构图里面，FM和DNN共享Dense Embeddings。这样FM和DNN的输入相比，少了数值特征。
> 在下面这篇文章的代码里，其把数值特征也映射到了4维的特征空间。
> https://zhuanlan.zhihu.com/p/84526966
> 请问，在DeepFM里面，对数值特征的处理，有什么思路或者技巧么？

  你好，我在代码中没有把数值特征加入FM是因为我这里是单纯地进行稀疏特征之间的交叉。但是按照论文中的结构图的话，其实FM那块也应该把数值特征加入其中，进行特征交叉。因为我觉得稠密（数值）特征和稀疏特征之间交叉的意义很模糊。
  比如，例子1：（时间=晚上，食物=泡面）的特征组合对ctr点击率会有直观的影响，而例子2：（时间=7.39423, 食物=泡面）的特征组合其实就不会有很大的影响，因为这里将时间单纯的指定到具体的值其实这样的组合太多了（考虑时间是连续值的话），但是你如果将时间分桶变成离散的类别特征其实就符合我例子1的情况了。所以在这里我没有将数值特征放入FM中进行交叉，不过你愿意的话是可以将数值特征加入到FM中，这从公式上理解是没有问题的。
  你说的文章的代码其实就是我刚刚说的将数值特征分桶，比如把连续的时间划分成{早上，中午，晚上，半夜}，变成one-hot变为4维特征，这也就可以理解为你问的对数值特征的处理技巧。
  感谢提问，有疑问的话可以继续追问😃

**Comment by timor1988** [comment]
又看了一些文章，发现DeepFM里面的dense embedding就是默认通过类别特征产生的。连续特征不经过embedding。王喆的《深度学习推荐系统》这本书里介绍的Deep Crossing（P61），类别特征经过了embedding，数值特征直接和embedding之后的类别特征拼接为一个长向量。和你代码中做了相同的处理。实际中的一个问题：发现这些模型的输入特征都是one-hot编码版本的，如果是用户关键词画像特征，就会使用multi-hot编码，DNN部分：一个field特征产生的embedding可以是一个field中多个为1的特征对应的嵌入矩阵的多行求和；FM部分：也是嵌入矩阵对应特征为1的行进行求和，一个field得到一个输出向量，所有field得到的输出拼接起来输入到FM()方法。但是这样做感觉就成了field之间进行了FM，而不是field具体的特征进行FM。

**Comment by HanpyBin** [comment]
> 又看了一些文章，发现DeepFM里面的dense embedding就是默认通过类别特征产生的。连续特征不经过embedding。王喆的《深度学习推荐系统》这本书里介绍的Deep Crossing（P61），类别特征经过了embedding，数值特征直接和embedding之后的类别特征拼接为一个长向量。和你代码中做了相同的处理。实际中的一个问题：发现这些模型的输入特征都是one-hot编码版本的，如果是用户关键词画像特征，就会使用multi-hot编码，DNN部分：一个field特征产生的embedding可以是一个field中多个为1的特征对应的嵌入矩阵的多行求和；FM部分：也是嵌入矩阵对应特征为1的行进行求和，一个field得到一个输出向量，所有field得到的输出拼接起来输入到FM()方法。但是这样做感觉就成了field之间进行了FM，而不是field具体的特征进行FM。

连续特征无法进行embedding，连续特征的embedding多为其权重的表示，即用维度为1的embedding代替fc层。关于你最后说的，确实是一个field一个embedding，如将历史购买记录用一条向量表示，这主要为了保证输入维度的统一性，不然的话，如果某人购买2个产品，而另一个人买了3个，则维度不统一，所以在这里采用求和亦或是其他方法也好，最终得到历史购买商品的embedding表示，进行FM也是利用这个embedding进行，而不是具体购买商品的embedding，具体可参考DIN

**Comment by timor1988** [comment]
额，明白了。
发现了最后一个问题，代码最终的输出为
`return torch.cat([output, 1-output], dim=1)`，
此时，outputs第1个值为点击的概率(=1的概率)，第2个值为不点击的概率（=0的概率）。
应该的顺序为：第1个值为不点击的概率(=0的概率)，第2个值为点击的概率（=1的概率），这样，在计算交叉熵的时候，y取1时候对应的就是点击的概率。
假设模型准确率100%，给定一个负样本，经过sigmoid之后的输出认为是0，则模型前向计算返回值应该是`outputs=torch.tensor([1-0=1.,0])`，此时`y=torch.tensor([0])`
二者交叉熵：
` 
loss = nn.CrossEntropyLoss()  
res=loss(outputs,y) # tensor(0.3133)
`
如果输出为``torch.tensor([0.,1-0=1])``，此时交叉熵为`tensor(1.3133)`

**Comment by HanpyBin** [comment]
> 额，明白了。
> 发现了最后一个问题，代码最终的输出为
> `return torch.cat([output, 1-output], dim=1)`，
> 此时，outputs第1个值为点击的概率(=1的概率)，第2个值为不点击的概率（=0的概率）。
> 应该的顺序为：第1个值为不点击的概率(=0的概率)，第2个值为点击的概率（=1的概率），这样，在计算交叉熵的时候，y取1时候对应的就是点击的概率。
> 假设模型准确率100%，给定一个负样本，经过sigmoid之后的输出认为是0，则模型前向计算返回值应该是`outputs=torch.tensor([1-0=1.,0])`，此时`y=torch.tensor([0])`
> 二者交叉熵：
> ` loss = nn.CrossEntropyLoss() res=loss(outputs,y) # tensor(0.3133)`
> 如果输出为`torch.tensor([0.,1-0=1])`，此时交叉熵为`tensor(1.3133)`

你这里主观带入了负样本经过sigmoid之后就是0，按照我这里的写法[output,1-output]的话，我的负样本经过sigmoid等于1而不是0, 那最后的输出是[1, 0], 而真正的y是0，此时交叉熵是tensor(0.3133)，不知道你能不能明白我意思😂

**Comment by timor1988** [comment]
> > 额，明白了。
> > 发现了最后一个问题，代码最终的输出为
> > `return torch.cat([output, 1-output], dim=1)`，
> > 此时，outputs第1个值为点击的概率(=1的概率)，第2个值为不点击的概率（=0的概率）。
> > 应该的顺序为：第1个值为不点击的概率(=0的概率)，第2个值为点击的概率（=1的概率），这样，在计算交叉熵的时候，y取1时候对应的就是点击的概率。
> > 假设模型准确率100%，给定一个负样本，经过sigmoid之后的输出认为是0，则模型前向计算返回值应该是`outputs=torch.tensor([1-0=1.,0])`，此时`y=torch.tensor([0])`
> > 二者交叉熵：
> > ` loss = nn.CrossEntropyLoss() res=loss(outputs,y) # tensor(0.3133)`
> > 如果输出为`torch.tensor([0.,1-0=1])`，此时交叉熵为`tensor(1.3133)`
> 
> 你这里主观带入了负样本经过sigmoid之后就是0，按照我这里的写法[output,1-output]的话，我的负样本经过sigmoid等于1而不是0, 那最后的输出是[1, 0], 而真正的y是0，此时交叉熵是tensor(0.3133)，不知道你能不能明白我意思😂

额。这样也没问题。就是sigmoid的输出，越靠近0，越为正样本。。但是一般的神经网络经过sigmoid之后的值属于(0,1)，越靠近1，越为正样本，而这个数据集里，正样本的y也等于1。

**Comment by HanpyBin** [comment]
> > > 额，明白了。
> > > 发现了最后一个问题，代码最终的输出为
> > > `return torch.cat([output, 1-output], dim=1)`，
> > > 此时，outputs第1个值为点击的概率(=1的概率)，第2个值为不点击的概率（=0的概率）。
> > > 应该的顺序为：第1个值为不点击的概率(=0的概率)，第2个值为点击的概率（=1的概率），这样，在计算交叉熵的时候，y取1时候对应的就是点击的概率。
> > > 假设模型准确率100%，给定一个负样本，经过sigmoid之后的输出认为是0，则模型前向计算返回值应该是`outputs=torch.tensor([1-0=1.,0])`，此时`y=torch.tensor([0])`
> > > 二者交叉熵：
> > > ` loss = nn.CrossEntropyLoss() res=loss(outputs,y) # tensor(0.3133)`
> > > 如果输出为`torch.tensor([0.,1-0=1])`，此时交叉熵为`tensor(1.3133)`
> > 
> > 
> > 你这里主观带入了负样本经过sigmoid之后就是0，按照我这里的写法[output,1-output]的话，我的负样本经过sigmoid等于1而不是0, 那最后的输出是[1, 0], 而真正的y是0，此时交叉熵是tensor(0.3133)，不知道你能不能明白我意思😂
> 
> 额。这样也没问题。就是sigmoid的输出，越靠近0，越为正样本。。但是一般的神经网络经过sigmoid之后的值属于(0,1)，越靠近1，越为正样本，而这个数据集里，正样本的y也等于1。

从习惯来说的话确实output放在右边😂

**Comment by namespace-io** [comment]
这里我理解 数值不进入FM的二阶计算的原因是不是可以看做和 一阶的计算结果 一并处理了，一阶的计算结果是wx + b, 比如说数值和数值的二阶计算可以看做合并加到一阶中的b里面 然后数值和离散的计算合并到wx中

**Comment by namespace-io** [comment]
> FM部分：也是嵌入矩阵对应特征为1的行进行求和，一个field得到一个输出向量，所有field得到的输出拼接起来输入到FM()方法。但是这样做感觉就成了field之间进行了FM，而不是field具体的特征进行FM

这里我理解就是 one-hot之后的向量进行了 FM 不过计算过程被简化了  因为其他的特征为0， 
![image](https://user-images.githubusercontent.com/29119076/196664468-cbefdd84-f3e7-47cb-a13a-9416a0a097f7.png)
